crates/primitives/src/macros.rs

```diff
@@ -0,0 +1,108 @@
+macro_rules! impl_buf {
+    ($name:ident, $len:expr) => {
+        impl $name {
+            pub const LEN: usize = $len;
+
+            pub fn new(data: [u8; $len]) -> Self {
+                Self(data)
+            }
+
+            pub fn as_slice(&self) -> &[u8] {
+                &self.0
+            }
+
+            pub fn as_mut_slice(&mut self) -> &mut [u8] {
+                &mut self.0
+            }
+        }
+
+        impl AsRef<[u8]> for $name {
+            fn as_ref(&self) -> &[u8] {
+                &self.0
+            }
+        }
+
+        impl AsMut<[u8]> for $name {
+            fn as_mut(&mut self) -> &mut [u8] {
+                &mut self.0
+            }
+        }
+
+        impl From<[u8; $len]> for $name {
+            fn from(data: [u8; $len]) -> Self {
+                Self(data)
+            }
+        }
+
+        impl From<$name> for [u8; $len] {
+            fn from(buf: $name) -> Self {
+                buf.0
+            }
+        }
+
+        impl<'a> From<&'a [u8; $len]> for $name {
+            fn from(data: &'a [u8; $len]) -> Self {
+                Self(*data)
+            }
+        }
+
+        impl Default for $name {
+            fn default() -> Self {
+                Self([0; $len])
+            }
+        }
+
+        // Add any other common impls here if needed.
+    };
```

Comment:
This is a good start, but we should add `Debug` and `Display` impls for these too.

Note this pattern to avoid allocs:

```rs

impl fmt::Debug for Buf20 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buf = [0; 40];
        hex::encode_to_slice(self.0, &mut buf).expect("buf: enc hex");
        f.write_str(unsafe { str::from_utf8_unchecked(&buf) })
    }
}

impl fmt::Display for Buf20 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut buf = [0; 6];
        hex::encode_to_slice(&self.0[..3], &mut buf).expect("buf: enc hex");
        f.write_str(unsafe { str::from_utf8_unchecked(&buf) })?;
        f.write_str("..")?;
        hex::encode_to_slice(&self.0[17..], &mut buf).expect("buf: enc hex");
        f.write_str(unsafe { str::from_utf8_unchecked(&buf) })?;
        Ok(())
    }
}
```